Optimize Span<T>.Fill implementation

GrabYourPitchforks opened the pull request:
This optimizes `Span<T>.Fill` via three primary mechanisms:

- For _T = byte_, forwards directly to the _initblk_ (_memset_) implementation
- For _T = \<primitive\>_, uses a SIMD-optimized worker loop if feasible
- Removes requirement for caller to stack-spill the span argument before calling main worker API

The central SIMD loop doesn't attempt to perform any type of alignment optimization. We can consider adding this in the future if benchmarking shows this to be a worthwhile addition.

I also didn't investigate any other call sites throughout the runtime + libraries to see if they should be migrated from whatever existing code they might have to this `Span<T>.Fill` implementation. That can come as a future commit to this PR or as a future PR.

<details>
<summary>Benchmark code</summary>

```cs
[GenericTypeArguments(typeof(byte))]
[GenericTypeArguments(typeof(char))]
[GenericTypeArguments(typeof(int))]
[GenericTypeArguments(typeof(long))]
[GenericTypeArguments(typeof(float))]
[GenericTypeArguments(typeof(double))]
[GenericTypeArguments(typeof(decimal))]
[GenericTypeArguments(typeof(string))]
public class SpanFillRunner<T>
{
    private T[] _arr;

    private T _value;

    [Params(0, 3, 7, 15, 16, 24, 128, 512)]
    public int Size;

    [GlobalSetup]
    public void Setup()
    {
        _arr = new T[Size];

        _value = (T)((IConvertible)42).ToType(typeof(T), null);
    }

    [Benchmark]
    public void Fill()
    {
        var arr = _arr;
        _ = arr.Length; // prove not null
        arr.AsSpan().Fill(_value);
    }
}
```
</details>

### Benchmark results:

<details>
<summary>byte</summary>

(Note: The internal _memset_ routine uses nontemporal stores, which could explain the blazing fast runtime.)

| Method |        Job | Toolchain | Size |     Mean |     Error |    StdDev | Ratio | RatioSD |
|------- |----------- |---------- |----- |---------:|----------:|----------:|------:|--------:|
|   **Fill** | **Job-NKOJQM** |      **main** |    **0** | **1.882 ns** | **0.0093 ns** | **0.0083 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    0 | 1.382 ns | 0.0320 ns | 0.0300 ns |  0.74 |    0.01 |
|        |            |           |      |          |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |    **3** | **3.449 ns** | **0.0172 ns** | **0.0144 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    3 | 1.865 ns | 0.0445 ns | 0.0395 ns |  0.54 |    0.01 |
|        |            |           |      |          |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |    **7** | **3.439 ns** | **0.0112 ns** | **0.0100 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    7 | 1.854 ns | 0.0195 ns | 0.0163 ns |  0.54 |    0.01 |
|        |            |           |      |          |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **15** | **3.386 ns** | **0.0115 ns** | **0.0102 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   15 | 1.857 ns | 0.0219 ns | 0.0183 ns |  0.55 |    0.01 |
|        |            |           |      |          |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **16** | **3.424 ns** | **0.0115 ns** | **0.0096 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   16 | 1.845 ns | 0.0181 ns | 0.0151 ns |  0.54 |    0.01 |
|        |            |           |      |          |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **24** | **3.211 ns** | **0.0173 ns** | **0.0153 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   24 | 1.918 ns | 0.0093 ns | 0.0087 ns |  0.60 |    0.00 |
|        |            |           |      |          |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |  **128** | **5.484 ns** | **0.0543 ns** | **0.0454 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |  128 | 4.321 ns | 0.1117 ns | 0.1602 ns |  0.79 |    0.03 |
|        |            |           |      |          |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |  **512** | **7.521 ns** | **0.0423 ns** | **0.0395 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |  512 | 6.451 ns | 0.0442 ns | 0.0345 ns |  0.86 |    0.01 |
</details>

<details>
<summary>char</summary>

| Method |        Job | Toolchain | Size |       Mean |     Error |    StdDev | Ratio | RatioSD |
|------- |----------- |---------- |----- |-----------:|----------:|----------:|------:|--------:|
|   **Fill** | **Job-NKOJQM** |      **main** |    **0** |   **1.880 ns** | **0.0569 ns** | **0.0532 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    0 |   1.355 ns | 0.0039 ns | 0.0036 ns |  0.72 |    0.02 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |    **3** |   **3.436 ns** | **0.0137 ns** | **0.0128 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    3 |   2.511 ns | 0.0167 ns | 0.0131 ns |  0.73 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |    **7** |   **4.199 ns** | **0.0524 ns** | **0.0490 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    7 |   2.300 ns | 0.0121 ns | 0.0113 ns |  0.55 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **15** |   **5.261 ns** | **0.0325 ns** | **0.0304 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   15 |   2.591 ns | 0.0353 ns | 0.0330 ns |  0.49 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **16** |   **5.263 ns** | **0.0192 ns** | **0.0170 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   16 |   2.516 ns | 0.0150 ns | 0.0140 ns |  0.48 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **24** |   **6.809 ns** | **0.0126 ns** | **0.0106 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   24 |   2.296 ns | 0.0105 ns | 0.0098 ns |  0.34 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |  **128** |  **30.955 ns** | **0.1179 ns** | **0.1103 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |  128 |   3.300 ns | 0.0942 ns | 0.1723 ns |  0.10 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |  **512** | **130.535 ns** | **1.7414 ns** | **1.5437 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |  512 |   7.081 ns | 0.1504 ns | 0.1407 ns |  0.05 |    0.00 |
</details>

<details>
<summary>int</summary>

| Method |        Job | Toolchain | Size |       Mean |     Error |    StdDev | Ratio | RatioSD |
|------- |----------- |---------- |----- |-----------:|----------:|----------:|------:|--------:|
|   **Fill** | **Job-NKOJQM** |      **main** |    **0** |   **1.603 ns** | **0.0630 ns** | **0.1183 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    0 |   1.720 ns | 0.1739 ns | 0.5127 ns |  0.98 |    0.23 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |    **3** |   **2.611 ns** | **0.0225 ns** | **0.0211 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    3 |   2.030 ns | 0.0258 ns | 0.0241 ns |  0.78 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |    **7** |   **2.638 ns** | **0.0194 ns** | **0.0181 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    7 |   2.062 ns | 0.0059 ns | 0.0055 ns |  0.78 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **15** |   **4.415 ns** | **0.0084 ns** | **0.0075 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   15 |   1.614 ns | 0.0045 ns | 0.0042 ns |  0.37 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **16** |   **5.104 ns** | **0.0169 ns** | **0.0149 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   16 |   2.060 ns | 0.0046 ns | 0.0041 ns |  0.40 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **24** |   **7.028 ns** | **0.0173 ns** | **0.0162 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   24 |   1.815 ns | 0.0157 ns | 0.0147 ns |  0.26 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |  **128** |  **29.413 ns** | **0.0861 ns** | **0.0805 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |  128 |   5.971 ns | 0.0835 ns | 0.1609 ns |  0.21 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |  **512** | **117.060 ns** | **0.8512 ns** | **0.7962 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |  512 |  17.274 ns | 0.0655 ns | 0.0512 ns |  0.15 |    0.00 |
</details>

<details>
<summary>long</summary>

| Method |        Job | Toolchain | Size |       Mean |     Error |    StdDev | Ratio | RatioSD |
|------- |----------- |---------- |----- |-----------:|----------:|----------:|------:|--------:|
|   **Fill** | **Job-NKOJQM** |      **main** |    **0** |   **1.367 ns** | **0.0170 ns** | **0.0159 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    0 |   1.151 ns | 0.0086 ns | 0.0080 ns |  0.84 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |    **3** |   **2.940 ns** | **0.0849 ns** | **0.0977 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    3 |   2.055 ns | 0.0052 ns | 0.0044 ns |  0.69 |    0.02 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |    **7** |   **2.635 ns** | **0.0131 ns** | **0.0122 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    7 |   1.598 ns | 0.0203 ns | 0.0189 ns |  0.61 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **15** |   **4.427 ns** | **0.0123 ns** | **0.0115 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   15 |   1.662 ns | 0.0103 ns | 0.0097 ns |  0.38 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **16** |   **5.116 ns** | **0.0095 ns** | **0.0079 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   16 |   2.526 ns | 0.0072 ns | 0.0067 ns |  0.49 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **24** |   **6.957 ns** | **0.0298 ns** | **0.0279 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   24 |   2.974 ns | 0.0103 ns | 0.0096 ns |  0.43 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |  **128** |  **29.398 ns** | **0.0854 ns** | **0.0799 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |  128 |  13.634 ns | 0.0477 ns | 0.0423 ns |  0.46 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |  **512** | **116.306 ns** | **0.1551 ns** | **0.1295 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |  512 |  60.719 ns | 0.1973 ns | 0.1845 ns |  0.52 |    0.00 |
</details>

<details>
<summary>float</summary>

| Method |        Job | Toolchain | Size |       Mean |     Error |    StdDev | Ratio | RatioSD |
|------- |----------- |---------- |----- |-----------:|----------:|----------:|------:|--------:|
|   **Fill** | **Job-NKOJQM** |      **main** |    **0** |   **1.367 ns** | **0.0189 ns** | **0.0177 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    0 |   1.141 ns | 0.0078 ns | 0.0073 ns |  0.83 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |    **3** |   **3.011 ns** | **0.0395 ns** | **0.0369 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    3 |   2.055 ns | 0.0073 ns | 0.0069 ns |  0.68 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |    **7** |   **2.742 ns** | **0.0432 ns** | **0.0404 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    7 |   2.054 ns | 0.0233 ns | 0.0218 ns |  0.75 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **15** |   **4.565 ns** | **0.1167 ns** | **0.1949 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   15 |   1.590 ns | 0.0297 ns | 0.0264 ns |  0.34 |    0.02 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **16** |   **4.693 ns** | **0.0448 ns** | **0.0419 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   16 |   2.078 ns | 0.0031 ns | 0.0027 ns |  0.44 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **24** |   **6.848 ns** | **0.0843 ns** | **0.0789 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   24 |   1.815 ns | 0.0172 ns | 0.0161 ns |  0.27 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |  **128** |  **29.045 ns** | **0.0866 ns** | **0.0810 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |  128 |   6.326 ns | 0.0459 ns | 0.0429 ns |  0.22 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |  **512** | **115.432 ns** | **0.4247 ns** | **0.3973 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |  512 |  32.200 ns | 0.1476 ns | 0.1381 ns |  0.28 |    0.00 |
</details>

<details>
<summary>double</summary>

| Method |        Job | Toolchain | Size |       Mean |     Error |    StdDev | Ratio | RatioSD |
|------- |----------- |---------- |----- |-----------:|----------:|----------:|------:|--------:|
|   **Fill** | **Job-NKOJQM** |      **main** |    **0** |   **1.358 ns** | **0.0233 ns** | **0.0207 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    0 |   1.131 ns | 0.0163 ns | 0.0153 ns |  0.83 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |    **3** |   **2.956 ns** | **0.0491 ns** | **0.0459 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    3 |   2.024 ns | 0.0366 ns | 0.0342 ns |  0.68 |    0.02 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |    **7** |   **2.931 ns** | **0.0346 ns** | **0.0323 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    7 |   1.625 ns | 0.0335 ns | 0.0314 ns |  0.55 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **15** |   **4.455 ns** | **0.0533 ns** | **0.0499 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   15 |   1.678 ns | 0.0621 ns | 0.0829 ns |  0.36 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **16** |   **5.098 ns** | **0.0513 ns** | **0.0480 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   16 |   2.508 ns | 0.0802 ns | 0.1150 ns |  0.51 |    0.02 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **24** |   **6.853 ns** | **0.0835 ns** | **0.0781 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   24 |   2.943 ns | 0.0378 ns | 0.0353 ns |  0.43 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |  **128** |  **28.694 ns** | **0.2794 ns** | **0.2613 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |  128 |  13.549 ns | 0.0357 ns | 0.0298 ns |  0.47 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |  **512** | **115.105 ns** | **1.1767 ns** | **1.1007 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |  512 |  60.911 ns | 0.7672 ns | 0.7177 ns |  0.53 |    0.01 |
</details>

<details>
<summary>decimal</summary>

(This exercises non-SIMD code paths.)

| Method |        Job | Toolchain | Size |       Mean |     Error |    StdDev | Ratio | RatioSD |
|------- |----------- |---------- |----- |-----------:|----------:|----------:|------:|--------:|
|   **Fill** | **Job-NKOJQM** |      **main** |    **0** |   **1.387 ns** | **0.0146 ns** | **0.0136 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    0 |   1.350 ns | 0.0227 ns | 0.0212 ns |  0.97 |    0.02 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |    **3** |   **3.848 ns** | **0.0300 ns** | **0.0281 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    3 |   2.878 ns | 0.0370 ns | 0.0346 ns |  0.75 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |    **7** |   **5.386 ns** | **0.0591 ns** | **0.0553 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |    7 |   4.827 ns | 0.0740 ns | 0.0692 ns |  0.90 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **15** |  **10.885 ns** | **0.0963 ns** | **0.0804 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   15 |  10.343 ns | 0.0469 ns | 0.0391 ns |  0.95 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **16** |  **11.545 ns** | **0.0925 ns** | **0.0866 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   16 |  10.919 ns | 0.0234 ns | 0.0207 ns |  0.95 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |   **24** |  **16.824 ns** | **0.0822 ns** | **0.0769 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |   24 |  16.434 ns | 0.0904 ns | 0.0846 ns |  0.98 |    0.01 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |  **128** |  **87.277 ns** | **0.2652 ns** | **0.2480 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |  128 |  86.800 ns | 0.1626 ns | 0.1358 ns |  0.99 |    0.00 |
|        |            |           |      |            |           |           |       |         |
|   **Fill** | **Job-NKOJQM** |      **main** |  **512** | **347.978 ns** | **0.8310 ns** | **0.7773 ns** |  **1.00** |    **0.00** |
|   Fill | Job-DYLJDS |  spanfill |  512 | 349.540 ns | 1.1889 ns | 1.1121 ns |  1.00 |    0.00 |
</details>

<details>
<summary>string</summary>

(This exercises non-SIMD code paths.)

| Method |        Job | Toolchain | Size |         Mean |     Error |    StdDev | Ratio |
|------- |----------- |---------- |----- |-------------:|----------:|----------:|------:|
|   **Fill** | **Job-NKOJQM** |      **main** |    **0** |     **9.361 ns** | **0.0327 ns** | **0.0306 ns** |  **1.00** |
|   Fill | Job-DYLJDS |  spanfill |    0 |    11.414 ns | 0.0994 ns | 0.0930 ns |  1.22 |
|        |            |           |      |              |           |           |       |
|   **Fill** | **Job-NKOJQM** |      **main** |    **3** |    **22.621 ns** | **0.1303 ns** | **0.1088 ns** |  **1.00** |
|   Fill | Job-DYLJDS |  spanfill |    3 |    15.888 ns | 0.0810 ns | 0.0758 ns |  0.70 |
|        |            |           |      |              |           |           |       |
|   **Fill** | **Job-NKOJQM** |      **main** |    **7** |    **37.180 ns** | **0.4089 ns** | **0.3825 ns** |  **1.00** |
|   Fill | Job-DYLJDS |  spanfill |    7 |    21.691 ns | 0.3115 ns | 0.2762 ns |  0.58 |
|        |            |           |      |              |           |           |       |
|   **Fill** | **Job-NKOJQM** |      **main** |   **15** |    **62.400 ns** | **0.4737 ns** | **0.4431 ns** |  **1.00** |
|   Fill | Job-DYLJDS |  spanfill |   15 |    32.190 ns | 0.0847 ns | 0.0792 ns |  0.52 |
|        |            |           |      |              |           |           |       |
|   **Fill** | **Job-NKOJQM** |      **main** |   **16** |    **66.787 ns** | **0.2268 ns** | **0.2121 ns** |  **1.00** |
|   Fill | Job-DYLJDS |  spanfill |   16 |    34.926 ns | 0.2742 ns | 0.2565 ns |  0.52 |
|        |            |           |      |              |           |           |       |
|   **Fill** | **Job-NKOJQM** |      **main** |   **24** |    **97.488 ns** | **0.2009 ns** | **0.1781 ns** |  **1.00** |
|   Fill | Job-DYLJDS |  spanfill |   24 |    46.627 ns | 0.1166 ns | 0.1034 ns |  0.48 |
|        |            |           |      |              |           |           |       |
|   **Fill** | **Job-NKOJQM** |      **main** |  **128** |   **448.163 ns** | **0.8712 ns** | **0.8149 ns** |  **1.00** |
|   Fill | Job-DYLJDS |  spanfill |  128 |   200.046 ns | 0.5331 ns | 0.4452 ns |  0.45 |
|        |            |           |      |              |           |           |       |
|   **Fill** | **Job-NKOJQM** |      **main** |  **512** | **1,744.556 ns** | **3.5163 ns** | **3.2891 ns** |  **1.00** |
|   Fill | Job-DYLJDS |  spanfill |  512 |   748.678 ns | 1.5795 ns | 1.4002 ns |  0.43 |
</details>

Resolves https://github.com/dotnet/runtime/issues/24806.
Resolves https://github.com/dotnet/runtime/issues/7049.

/cc @carlossanlop @jozkee

GrabYourPitchforks committed:
Span.Fill perf improvements

GrabYourPitchforks commented:
_System.Linq.Expressions.Tests_ failures are known issue https://github.com/dotnet/runtime/issues/51346.

Tornhoof commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:
```diff
@@ -2,13 +2,160 @@
+                return; // nothing to do
+            }
+
+            if (typeof(T) != typeof(float) && typeof(T) != typeof(double) && !RuntimeHelpers.IsBitwiseEquatable<T>())
```
So half is not optimized? Maybe add it to your Benchmarks anyway?

tannergooding replied:
I'd definitely appreciate us adding `Half` support here. We'll be adding supporting to the vector types and elsewhere eventually.

It might even be nice if we had some assert we could put in places like this to help ensure "primitive" types are handled. Off the top of my head these are `byte`, `sbyte`, short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, and `double`.
There is also, as of recent C# versions, `nint` and `nuint`; and `Half` as of the latest .NET versions.

If you look a bit beyond that for other "primitives" or language special types (not just those that are directly primitive and have hardware acceleration for SIMD scenarios), there is also `char`, `bool`, `decimal`, and maybe a few others.

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:
```diff
@@ -2,13 +2,160 @@
+                {
+                    do
+                    {
+                        Unsafe.WriteUnaligned(ref Unsafe.AddByteOffset(ref Unsafe.As<T, byte>(ref refData), offset), vector);
```
Want to hoist these `ref Unsafe.As<T, byte>(ref refData)` in a `ref byte byteRef = ref Unsafe.As<T, byte>(ref refData);` and use `ref byteRef` so it's a bit nicer to read?

Codegen will be the same.

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:
```diff
@@ -2,13 +2,160 @@
+                // entry, but that's ok since we're splatting the same value for all entries, so this
+                // won't result in data corruption.
+
+                if ((totalByteLength & (nuint)(Vector<byte>.Count - 1)) != 0)
```
I believe you can omit this branch by just writing the vector at the end (idempotence, as you commented) always.

The check in L30 guarantees that at least one vector can be written. So just write it, and avoid this branch (with potential for mis-prediction).

The subtraction `totalByteLength - (uint)Vector<byte>.Count)` will be done at the address-resolution stage, so always writing should be cheaper than the branch.

stephentoub commented on src/libraries/System.Memory/tests/Span/Fill.cs:
```diff
@@ -147,5 +150,53 @@ public static unsafe void FillNativeBytes()
+                .Invoke(null, BindingFlags.DoNotWrapExceptions, null, new object[] { value }, null);
+        }
+
+        private static void FillWithRecognizedType_RunTest<T>(T value)
```
Maybe I'm misremembering, but doesn't xunit let you put [InlineData] with different types directly onto a generic test method, and it handles doing the relevant reflection/casting for you?

GrabYourPitchforks replied:
Looks like it's not supported. See https://github.com/xunit/xunit/issues/1378.

stephentoub replied:
I'm surprised, given we have tests like these:
https://github.com/dotnet/runtime/blob/907fc3e13a53820329890e59d8ff65eb286cca30/src/libraries/System.Collections/tests/Generic/Comparers/Comparer.Tests.cs#L14-L21
https://github.com/dotnet/runtime/blob/907fc3e13a53820329890e59d8ff65eb286cca30/src/libraries/System.Drawing.Common/tests/PenTests.cs#L23-L25
https://github.com/dotnet/runtime/blob/907fc3e13a53820329890e59d8ff65eb286cca30/src/libraries/System.Linq/tests/GroupByTests.DebuggerAttributes.cs#L13-L15

GrabYourPitchforks replied:
Huh, for some reason xunit always kept passing _T = object_ in for this unit test when I tried doing this. I'll try tweaking the signature to see if this fixes things.

GrabYourPitchforks replied:
Setting up the theory to support this ended up making the code look bad. I simplified a bit to put all the data inline and have it delegate to a common test helper. This avoids the reflection call and looks cleaner overall.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:
```diff
@@ -2,13 +2,160 @@
+                // There are [ 0, 2 * sizeof(Vector) ) bytes remaining.
+                // If there are >= sizeof(Vector) bytes remaining, write one vector now.
+
+                if ((totalByteLength & (nuint)Vector<byte>.Count) != 0)
```
This is subtle; I had to stare at this for a minute to convince myself it was correct.  Could you add a little more explanation to the comment above?

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:
```diff
@@ -2,13 +2,160 @@
+
+            {
+                nuint i = 0;
+                nuint stopLoopAtOffset = numElements & ~(uint)7;
```
Nit: can you move this down to just before the for loop below, since it only applies to that loop rather than to this whole cannot vectorize section?

stephentoub reviewed (approved):
Nice!

tannergooding commented on src/libraries/System.Memory/tests/Span/Fill.cs:
```diff
@@ -147,5 +150,53 @@ public static unsafe void FillNativeBytes()
+                yield return new object[] { typeof(float), 1.0f };
+                yield return new object[] { typeof(double), 1.0d };
+                yield return new object[] { typeof(string), "Hello world!" }; // shouldn't go down SIMD path
+                yield return new object[] { typeof(decimal), 1.0m }; // shouldn't go down SIMD path
```
Why not? `decimal` is a blittable value type that is 128-bits in length, it (like `Guid`) seems like a perfect candidate (outside the possibility of it being "rare")

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:
```diff
@@ -2,13 +2,160 @@
+                    {
+                        vector = (Vector<byte>)(new Vector<ulong>(Unsafe.As<T, ulong>(ref tmp)));
+                    }
+                    else
```
This seems like it should handle `Unsafe.SizeOf<T>() == 16`, it could at least handle `Vector2`, `Vector4`, `Vector128<T>`, `Guid`, and `decimal`, but potentially other types as well.

GrabYourPitchforks replied:
How do I get an arbitrary 16-byte T into a `Vector<T>`? There doesn't seem to be any relevant ctor or cast I can perform.

tannergooding replied:
It would basically be:
```csharp
var v128 = Unsafe.As<T, Vector128<byte>>(ref value);

if (sizeof(Vector<T>) == 32)
{
    var v256 = Vector256.Create(v128, v128);
    vector = v256.AsVector();
}
else
{
    Debug.Assert(sizeof(Vector<T>) == 16);
    vector = v128.AsVector();
}
```

SingleAccretion commented:
A note on the `memset` flavor used by the runtime: it is quite sensitive to alignment, at least for `x64`. [There is a comment](https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/coreclr/vm/comutilnative.cpp#L626) about it in runtime source.

GrabYourPitchforks commented:
@SingleAccretion good observation. I stepped through the memset implementation (used by initblk) and noticed that it performed alignment fixup before entering the main loop. So I wonder if this comment is stale?

SingleAccretion commented:
> So I wonder if this comment is stale?

Very possible. My comment was the result of me measuring this a few months back, and seeing the expected 2x regression (Ivy Bridge). If it is no longer accurate, that's great!

GrabYourPitchforks committed:
PR feedback, slight perf tweaks

GrabYourPitchforks committed:
Fix > to >=

GrabYourPitchforks commented:
Notable changes in latest commit:

- The code now allows any non-ref-containing type that's a power of 2 and fits into a vector, including _Half_, _Guid_, _decimal_, etc.
- Cleaned up some of the write logic per feedback from @gfoidl
- Cleaned up unit test logic to avoid reflection
- Dropped the "zero elements?" early-exit check at the beginning of the method, as we generally shouldn't optimize for uncommon scenarios, and it'll eventually skip all the branches and hit method exit soon enough anyway
- Changed the `SpanHelpers` routine to take _nuint_ instead of _uint_ since it's a cheap enough change and it makes this code path more resilient to any future `NativeSpan` additions

No major changes to the overall design behind the logic.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:
```diff
@@ -4,103 +4,119 @@
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>()) { goto CannotVectorize; }
+            if (!Vector.IsHardwareAccelerated) { goto CannotVectorize; }
+            if (Unsafe.SizeOf<T>() > Vector<byte>.Count) { goto CannotVectorize; }
+            if ((Unsafe.SizeOf<T>() & (Unsafe.SizeOf<T>() - 1)) != 0) { goto CannotVectorize; } // power of 2 check
```
We have `BitOperations.IsPow2` now

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:
```diff
@@ -4,103 +4,119 @@
                         goto CannotVectorize;
                     }
                 }
+                else if (Unsafe.SizeOf<T>() == 32)
```
This needs to check against `Vector<byte>.Count` as well

GrabYourPitchforks committed:
PR feedback

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs:
```diff
@@ -71,7 +72,15 @@ public static void Fill<T>(ref T refData, nuint numElements, T value)
+                    }
+                    else
+                    {
+                        Debug.Fail("Vector<T> isn't 256 bits in size?");
```
This shouldn't be a failure (unless I'm missing some other condition somewhere). It is valid to have a `Vector256<T>` on say ARM, where `Vector<T>` is only ever `16-bytes`.

It's just that the given `Vector256<T>` will have no hardware acceleration and will be entirely software driven.

tannergooding replied:
Likewise on a machine with AVX support but no AVX2 support (partial hardware acceleration for `V256`) or a machine without any AVX support (no hardware acceleration for `V256`)

GrabYourPitchforks replied:
There's already a condition at the top of the method which skips the vectorized code path if `Unsafe.SizeOf<T>() > Vector<byte>.Count`. This check is intended to catch `Vector<T>` unexpectedly growing to - say - 512 bits.

tannergooding replied:
Thanks, that would be the "I'm missing some other condition somewhere" case 😄

GrabYourPitchforks committed:
Work around initblk null check on mono

GrabYourPitchforks commented:
Notable change in latest commit:

Mono interpreter's implementation of initblk performs a null check on the incoming address before delegating to _memset_. Restoring the "don't call initblk for empty spans" logic seemed like a more surgical solution than trying to change the interpreter's initblk logic, which might have unintentional ripple effects. I'm only restoring this logic for mono, as coreclr's initblk logic handles the zero-length case just fine, and a zero-length input should be very rare.

GrabYourPitchforks commented:
Thanks all for the feedback! If there are any additional comments we can send a cleanup PR.

davidfowl commented on src/libraries/System.Private.CoreLib/src/System/Span.cs:
```diff
@@ -284,11 +284,18 @@ public void Fill(T value)
-                Unsafe.InitBlockUnaligned(ref Unsafe.As<T, byte>(ref _pointer.Value), Unsafe.As<T, byte>(ref value), (uint)_length);
+#if MONO
+                // Mono runtime's implementation of initblk performs a null check on the address.
+                // We'll perform a length check here to avoid passing a null address in the empty span case.
```
Is that a compatibility thing? Should we change it?

GrabYourPitchforks replied:
See comment at https://github.com/dotnet/runtime/pull/51365#issuecomment-821765770. I think I know where to make the change, but I was concerned that it would have broader implications and that I'd risk inadvertently breaking some other component. Putting the check here seemed the safest option. (This code had a check originally, so it's not a behavioral change from before this PR.)

davidfowl replied:
File an issue? Making the runtimes behave the same would be best if possible.

SingleAccretion replied:
Already filed - #51411 😄.

davidfowl replied:
Is that the same bug?
